server/internal/mcp/rpc_tools_call.go

```diff
@@ -89,8 +91,33 @@ func handleToolsCall(
 		return nil, oops.E(oops.CodeNotFound, errors.New("tool not found"), "tool not found").Log(ctx, logger)
 	}
 
+	toolURN, err := conv.GetToolURN(*tool)
+	if err != nil {
+		return nil, oops.E(oops.CodeUnexpected, err, "failed to get tool urn").Log(ctx, logger)
+	}
+
 	ciEnv := gateway.NewCaseInsensitiveEnv()
 
+	envRepo := repo.New(db)
+	sourceEnv, err := envRepo.GetEnvironmentForSource(ctx, repo.GetEnvironmentForSourceParams{
+		SourceKind: string(toolURN.Kind),
+		SourceSlug: toolURN.Source,
+	})
+	if err != nil && !errors.Is(err, pgx.ErrNoRows) {
+		return nil, oops.E(oops.CodeUnexpected, err, "failed to get environment from source").Log(ctx, logger)
+	}
+
+	if err == nil {
+		sourceEnvVars, err := env.Load(ctx, payload.projectID, gateway.ID(sourceEnv.ID))
+		if err != nil && !errors.Is(err, gateway.ErrNotFound) {
+			return nil, oops.E(oops.CodeUnexpected, err, "failed to load source environment variables").Log(ctx, logger)
+		}
+
+		for k, v := range sourceEnvVars {
+			ciEnv.Set(k, v)
+		}
+	}
+
 	// IMPORTANT: MCP servers accessed in a public manner or not gram authenticated, there is no concept of using stored environments for them
 	if envSlug != "" && payload.authenticated {
 		storedEnvVars, err := env.Load(ctx, payload.projectID, gateway.Slug(envSlug))
```

Comment:
It feels like this is where your code should hook in. Like we should only ever be loading one environment (I think?) so we either need to load the attached one or the one requested by the envSlug

Reply:
This is more or less how the delineation between user and system configuration works in this system. The idea is that if environment is specified by slug that's a means of using gram environments to specify user configuration, which might be entirely disjoint from system configuration